### DISCUSSION TITLE
Typescript

### DISCUSSION THREAD

**Discussion by akmaluddin** [adobe/react-spectrum#5021] [Q&A] [answered]
I am wondering is there a full example of working typescript

I keep bumping into type issues that seems to convoluted to fix and most working examples and docs doesn't fully document the types or even worse, the running examples are from a previous version that doesn't match to the latest

Would love to have this implemented but now its just too difficult and I am not comfortable in using `any` type just as an escape hatch

**Comment by snowystinger** [comment]
Do you have a specific issue that you're running into? All of our examples on our docs have as much TS as they need in order to be functional. It is definitely minimal and we do have a goal to expand them. https://github.com/adobe/react-spectrum/issues/865

Can you point to an example that is running on a previous version?

**Comment by akmaluddin** [comment]
Sure! I am currently following the examples of Combobox found [here](https://react-spectrum.adobe.com/react-aria/useComboBox.html#usecombobox), the code examples have very minimal type reference so its very difficult to nail the right types.

So I head to the running examples [here](https://codesandbox.io/s/hardcore-moon-xzc4r?file=/src/Popover.tsx) and noticed that the type imports are different as well as the package versions

e.g. in the running example 
```jsx
import type { AriaPopoverProps } from "@react-aria/overlays";
``` 

should be I guess
```jsx
import type { AriaPopoverProps } from "react-aria";
```

**Reply by snowystinger** [reply]
Doesn't matter which one you use, it's just a re-export. https://github.com/adobe/react-spectrum/blob/main/packages/react-aria/src/index.ts#L67 We do that for ease of use, so you don't have to explicitly add all of our packages to your dependencies, just the relevant monopackage.

I forked and updated the versions on that codesandbox, I'm not seeing any issues, should I be?

I'll let the owner of that sandbox know they should update the versions.

**Reply by akmaluddin** [reply]
Anything with `@react-aria` will throw `Cannot find module @react-aria`

**Reply by akmaluddin** [reply]
I am running `react-aria: 3.27.0` btw, sorry should have included this

**Reply by snowystinger** [reply]
Huh... I don't see any errors in the codesandbox. Here's one with the versions bumped to the same as you've indicated https://codesandbox.io/s/hopeful-silence-hydq6z?file=/src/App.tsx

**Reply by akmaluddin** [reply]
Hmmm, this is what I am getting after `tsc`

```
> tsc

src/search-box/search-box.tsx:7:39 - error TS2307: Cannot find module '@react-aria/overlays' or its corresponding type declarations.

7 import type { AriaPopoverProps } from '@react-aria/overlays';
```

for brevity, here are what I might think are salient packages installed as deps

```json
"dependencies": {
        "react": "^18.2.0",
        "@types/react": "^18.2.14",
        "react-aria": "^3.27.0",
        "react-aria-components": "1.0.0-alpha.6",
        "react-icons": "^4.7.1",
        "react-stately": "^3.25.0"
    },
```

**Reply by snowystinger** [reply, marked as answer]
I think if you haven't declared an explicit dependency on `@react-aria/overlays` it's better to import it from `react-aria`. That way you aren't relying on it being installed via some implicit dependency. Maybe that's why your TS is complaining?

**Reply by akmaluddin** [reply]
It could be, but somehow in the code sandbox you provided that behaviour is not being observed. 🤷‍♂️. JS can be weird

**Reply by hastebrot** [reply]
I use this `.npmrc` so vscode/tsc finds the files:

```
# let pnpm include @react-aria/* and @react-stately/* in node_modules.
public-hoist-pattern[]=*react-aria*
public-hoist-pattern[]=*react-stately*
```

works with npm and pnpm. bun automatically does this.

**Comment by akmaluddin** [comment]
Also just to double check with the code examples

```jsx
import {Item, useComboBoxState} from 'react-stately';
```

Is Item an export from react-stately or RAC?

**Reply by snowystinger** [reply]
The answer is, it depends. If you're using RAC for the component rendering the collection, then you should use `Item` from RAC. Eventually, everything will be using that Item/collection implementation, but we haven't gotten there yet. You could use the one from RAC, but you'd need to do a lot of sluething through source code for how to use it for now.

**Reply by akmaluddin** [reply]
Understood! But when I search the docs, `Item` is not documented under react-stately so I am not sure what props are configured for `Item`

**Reply by snowystinger** [reply]
If you're using the Item from stately, then the Item props are https://github.com/adobe/react-spectrum/blob/cea5511d1419350f2e81c1f1bdb0e7134f25d97d/packages/%40react-types/shared/src/collections.d.ts#L15

I think usually we expect ItemProps to be inferred though, what's the use case for needing explicit access to them? Could help determine where to put the reference in our docs if we decide to do that.

**Reply by akmaluddin** [reply]
Well from my perspective, I have no reference to the types unless I looked through the source code. Say I am following the example of the [useComboBox](https://react-spectrum.adobe.com/react-aria/useComboBox.html)

The example provided

```tsx
...
import {Item, useComboBoxState} from 'react-stately';
...
  <Item key="red panda">Red Panda</Item>
  <Item key="cat">Cat</Item>
  <Item key="dog">Dog</Item>
  <Item key="aardvark">Aardvark</Item>
  <Item key="kangaroo">Kangaroo</Item>
  <Item key="snake">Snake</Item>
```

And if I were to find what props are part of `Item` and what they do, I only get docs for RAC rather then stately. It might be useful to document the interface of Item so that I can understand their usage :) 

E.g.

If I could be rendering a ReactElement as the children, having textValue props could be useful to populate the combobox

**Reply by snowystinger** [reply]
Yeah, you do have to click through a little, we do have them documented. If you follow the link in the Dynamic Collection section over to https://react-spectrum.adobe.com/react-stately/collections.html#static-collections, at the bottom of this section, we have an API popover for Item.
Is that what you're asking for to be improved? I agree this is non-obvious. Does your editor take you to the right declaration if you CMD+click on an `Item`? (shortcut key may change depending on OS and Editor)
<img width="740" alt="Screenshot 2023-09-06 at 5 12 13 PM" src="https://github.com/adobe/react-spectrum/assets/698229/0d755b89-0eb7-49c5-9251-e710321d11e2">

**Reply by akmaluddin** [reply]
Ahh, that was not apparent for me 😅!. Yep I can follow through the definitions through my IDE and can quickly identify the types but would be nice if its documented clearly under `react-stately` as I am not so sure if its stable to be used.

Also, I am in the midst of introducing `react-aria` into the codebase, and some of the junior developers will have trouble understanding what props are part of it.

**Reply by snowystinger** [reply]
No worries, it is quite stable. I think we can definitely feature it a little more prominently somewhere. The trick will just be where.

**Reply by akmaluddin** [reply]
Somewhere in react stately? Since its imported from `react-stately`. This is just my opinion 😁
<img width="335" alt="Screenshot 2023-09-06 at 10 39 30" src="https://github.com/adobe/react-spectrum/assets/16938323/8fe61f32-2de6-4bff-b88c-5c93eca0151c">

**Reply by snowystinger** [reply]
I've opened an issue https://github.com/adobe/react-spectrum/issues/5038
